### PR TITLE
fix DoH timeout

### PR DESCRIPTION
**Pull request by arm64v8a**
See #289
And also https://github.com/SagerNet/SagerNet/issues/129#issuecomment-855656860

This timeout happens when mux is disabled or not supported (socks outbound).

https://github.com/v2fly/v2ray-core/blob/b922eb7779ed89f5cdf1bfa14155e95098805415/app/dns/nameserver_doh.go#L237-L238

Tested config:
```
{
    "log": {
        "loglevel": "info"
    },
    "dns": {
        "servers": [
            "https://1.1.1.1/dns-query"
        ]
    },
    "stats": {},
    "inbounds": [
        {
            "port": 114514,
            "listen": "127.0.0.1",
            "protocol": "socks",
            "settings": {
                "auth": "noauth",
                "udp": true
            }
        }
    ],
    "outbounds": [
        {
            "protocol": "socks",
            "settings": {
                "servers": [
                    {
                        "address": "127.0.0.1", // Maybe a naiveproxy
                        "port": 1919810,
                        "users": []
                    }
                ]
            }
        },
        {
            "tag": "direct",
            "protocol": "freedom",
            "settings": {}
        },
        {
            "tag": "blocked",
            "protocol": "blackhole",
            "settings": {}
        }
    ],
    "routing": {
        "domainStrategy": "IPIfNonMatch",
        "rules": [
            {
                "type": "field",
                "ip": [
                    "geoip:private"
                ],
                "outboundTag": "blocked"
            },
            {
                "type": "field",
                "ip": [
                    "geoip:cn"
                ],
                "outboundTag": "direct"
            },
            {
                "type": "field",
                "domain": [
                    "geosite:cn"
                ],
                "outboundTag": "direct"
            }
        ]
    }
}
```
Vmess/VLESS is not tested because I don't have these proxies.

### CODE REVIEW

**Comment by rurirei** [comment]
> Vmess/VLESS is not tested because I don't have these proxies.

you could build a local server to test those.

**Comment by arm64v8a** [comment]
Vmess/VLESS are tested. (mux enabled/disabled)

**Comment by boypt** [comment]
I don't see this PR changes any thing: `defer cancel()` is called right after the function is returned.

File: app/dns/nameserver_doh.go [hunk 1]
```diff
@@ -249,6 +247,7 @@ func (s *DoHNameServer) sendQuery(ctx context.Context, domain string, clientIP n
 			}
 			resp, err := s.dohHTTPSContext(dnsCtx, b.Bytes())
 			if err != nil {
+				cancel()
```
Comment (vcptr) [inline app/dns/nameserver_doh.go:250]: I don't see this changes any thing: defer cancel() is called right after the function is returned.

Comment (Loyalsoldier) [reply app/dns/nameserver_doh.go:250]: Actually something is different.

Normally, there are two DNS requests, one is in A type, the other is in AAAA type. Both should get results.

In the old code, if one request returns without error and the other will be cancelled, that's not what we want.

Comment (rurirei) [reply app/dns/nameserver_doh.go:250]: but `dnsCtx` is the copied context for one dns request (for func(r *Request)), where i'm missed? @Loyalsoldier

Comment (arm64v8a) [reply app/dns/nameserver_doh.go:250]: @vcptr
`defer cancel()` is called right after the function is returned, including:
1. Successfully got answer
2. Error

I know that `canel() `do not impact the parent context. Maybe HTTP client did something when `dnsCtx` is cancelled, resulting: `v2ray.com/core/proxy/socks: connection ends > context canceled`

So I make the `cancel() `be called only when HTTP request fails. This is the change.

---

@rurirei 
I tried to revert some commit, but even `dnsCtx := context.Background()` timeout as usual.

The causes may be hidden in the `net/http` as I said above.

If `defer cancel()` is necessary, we need a better way to fix this.

First of all, I don't know why HTTP client keep alive requires mux support.

https://github.com/v2fly/v2ray-core/blob/b922eb7779ed89f5cdf1bfa14155e95098805415/app/dns/nameserver_doh.go#L49-L55

As a contrast, DoHLocalNameServer has keep alive as a Go HTTP client should.

Comment (vcptr) [reply app/dns/nameserver_doh.go:250]: OK I got yr point.
The change is to **not to call `cancel` on success results**.

I think the DOH layer http request should use an independent context, cancelling the DOH deadline context should not interrupt the lower layer context.

I have been away from looking deep into v2ray source for quite a while, please verify that the `dnsCtx` is not used into proxy.

But, cancelling the context whenever finished is necessary, [according to manual](https://golang.org/pkg/context/#WithDeadline)   , this PR could cause resource leak. 

> Canceling this context releases resources associated with it, so code should call cancel as soon as the operations running in this Context complete.

Comment (arm64v8a) [reply app/dns/nameserver_doh.go:250]: @vcptr 

> I tried to revert some commit, but even dnsCtx := context.Background() timeout as usual.

And I tried to revert more: v4.22.0 released in Dec 2019 doesn't timeout.

Your first version was good.

> I have been away from looking deep into v2ray source for quite a while, please verify that the dnsCtx is not used into proxy.

I don't have a deep understanding of v2ray's design, so I afraid I can't verify that.

Comment (rurirei) [reply app/dns/nameserver_doh.go:250]: in fact i cannot find the problem of dnsCtx, that's just a local variable for one request. buildMsg() makes two requests for type A and AAAA separately. @arm64v8a

**Comment by rurirei** [comment]
@arm64v8a need the logs of both type A and AAAA request.
```jsonc
    "dns": {
        ...,
       "queryStrategy": "useIP"
    }
```

**Comment by arm64v8a** [comment]
UseIP(default) for V2ray v4.39.2
```
2021/06/07 14:59:30 [Info] [1932474829] v2ray.com/core/proxy/socks: TCP Connect request to tcp:www.google.com:443
2021/06/07 14:59:30 [Info] v2ray.com/core/app/dns: DOH//1.1.1.1 querying: www.google.com.
2021/06/07 14:59:30 [Info] v2ray.com/core/app/dispatcher: default route for tcp:1.1.1.1:443
2021/06/07 14:59:30 [Info] v2ray.com/core/transport/internet/tcp: dialing TCP to tcp:127.0.0.1:naive
2021/06/07 14:59:30 [Info] v2ray.com/core/app/dispatcher: default route for tcp:1.1.1.1:443
2021/06/07 14:59:30 [Info] v2ray.com/core/transport/internet/tcp: dialing TCP to tcp:127.0.0.1:naive
2021/06/07 14:59:30 [Info] v2ray.com/core/app/dns: DOH//1.1.1.1 got answer: www.google.com. TypeAAAA -> [xxxxxxxxxx] 114.514ms
2021/06/07 14:59:30 [Info] v2ray.com/core/app/proxyman/outbound: failed to process outbound traffic > v2ray.com/core/proxy/socks: connection ends > context canceled
2021/06/07 14:59:30 [Error] v2ray.com/core/app/dns: failed to retrieve response > Post "https://1.1.1.1/dns-query": io: read/write on closed pipe
2021/06/07 14:59:34 [Info] [1932474829] v2ray.com/core/app/dispatcher: default route for tcp:www.google.com:443
2021/06/07 14:59:34 [Info] [1932474829] v2ray.com/core/transport/internet/tcp: dialing TCP to tcp:127.0.0.1:naive
```

UseIP for patched v4.39.2 or V2ray v4.22.0

```
2021/06/09 18:57:01 [Info] [244328656] proxy/socks: TCP Connect request to tcp:www.google.com:443
2021/06/09 18:57:01 [Info] app/dns: DOH//1.1.1.1 querying: www.google.com.
2021/06/09 18:57:01 [Info] app/dispatcher: default route for tcp:1.1.1.1:443
2021/06/09 18:57:01 [Info] transport/internet/tcp: dialing TCP to tcp:127.0.0.1:naive
2021/06/09 18:57:01 [Info] app/dispatcher: default route for tcp:1.1.1.1:443
2021/06/09 18:57:01 [Info] transport/internet/tcp: dialing TCP to tcp:127.0.0.1:naive
2021/06/09 18:57:02 [Info] app/dns: DOH//1.1.1.1 got answer: www.google.com. TypeA -> [******] 114.514ms
2021/06/09 18:57:02 [Info] app/dns: DOH//1.1.1.1 got answer: www.google.com. TypeAAAA -> [******] 114.514ms
2021/06/09 18:57:02 [Info] [244328656] app/dispatcher: default route for tcp:www.google.com:443
2021/06/09 18:57:02 [Info] [244328656] transport/internet/tcp: dialing TCP to tcp:127.0.0.1:naive
2021/06/09 18:57:02 tcp:127.0.0.1:44564 accepted tcp:www.google.com:443
```

You can see that A query fails after AAAA is done in v4.39.2

**Comment by rurirei** [comment]
thanks for your logs. i cannot find the problem of dnsCtx, why it' that?

**Comment by rurirei** [comment]
could you try https://github.com/rurirei/v2ray-core/commit/49116178758ad89f30b9e51865a6710c127de96e

**Comment by arm64v8a** [comment]
@rurirei  your fix doesn't work. log is simlar to v4.39.2.

Let's keep this PR Open until I/somebody find out a better solution.

**Comment by rurirei** [comment]
@arm64v8a https://github.com/rurirei/v2ray-core/commit/5db124909cb143c0692b45ee54975d93e59085c5 should work, doing the same thing as your pr.

c.c. @xiaokangwang could you help find the problem of dnsCtx?

**Comment by arm64v8a** [comment]
Unable to fix / add keep-alive.

**Pull request closed without merging**